torch/distributed/optim/zero_redundancy_optimizer.py

```diff
@@ -547,40 +543,41 @@ def _setup_flat_buffers(self) -> None:
-            if device not in self.buckets.keys():
-                self.buckets[device] = []
-
-            # Make parameters a view of the bucket
-            for dst_rank, params in enumerate(per_rank_params):
-                if len(params) > 0:
-
-                    # Clone the non-trainable params, if in a bucket it will get destroyed
-                    for param in filter(lambda x: not x.requires_grad, params):
+        for rank, param_groups in enumerate(self.partition_parameters()):
+            # Clone the non-trainable params, find the buffer size and dtype
+            # for the trainable params' bucket, and compile a list of the
+            # trainable params
+            buffer_size = 0
+            dtype = None
+            trainable_params = []
+            for param_group in param_groups:
+                for param in param_group["params"]:
+                    if (not _is_trainable(param)):
                         param.data = param.data.detach().clone()
-
-                    # Merge all the trainable params in a single bucket
-                    trainable_params = list(filter(_is_trainable, params))
-                    buffer_size = sum(map(lambda x: x.numel(), trainable_params))
-                    bucket = torch.empty(buffer_size, dtype=params[0].dtype, device=device)
-                    offset = 0
-
-                    for param in trainable_params:
-                        offset_next = offset + param.numel()
-                        bucket[offset:offset_next].copy_(param.data.flatten())
-                        param.data = bucket[offset:offset_next].view_as(param.data)
-                        offset = offset_next
-
-                    # Either replace the existing bucket, or create it
-                    if len(self.buckets[device]) == dst_rank:
-                        self.buckets[device].append(bucket)
                     else:
-                        self.buckets[device][dst_rank] = bucket
-                else:
-                    self.buckets[device].append(torch.zeros(1, device=device))
+                        buffer_size += param.numel()
+                        trainable_params.append(param)
+                    dtype = param.dtype
+
+            # Create a dummy bucket if there are no params
+            if (buffer_size == 0):
+                self.buckets.append(torch.zeros(1, device=self._device))
+                continue
+
+            # Otherwise, construct the bucket
+            bucket = torch.empty(buffer_size, dtype=dtype, device=self._device)
+            offset = 0
+            for param in trainable_params:
+                offset_next = offset + param.numel()
+                bucket[offset:offset_next].copy_(param.data.flatten())
+                param.data = bucket[offset:offset_next].view_as(param.data)
+                offset = offset_next
+
+            # Either replace the existing bucket or create it
```

Comment:
I did not quite follow the reasoning behind the need for this branching. Although, I replicated the logic for consistency. I am curious why we do not need the same branching for the dummy bucket case. In other words, why do we _always_ append in that case and not have to consider overriding the existing bucket?

Reply:
I don't recall the context here. One reason might be that `_setup_flat_buffers` can be called multiple times, and only the first time it should append, and all subsequent invocations should just override. @blefaudeux do you remember the motivation behind this?